### PR TITLE
KFP - Replaces the guide to building components and pipelines

### PR DESCRIPTION
**Pull request by joeliedtke**
Adds a new guide to building a pipeline that replaces the guide to building components and pipelines.

- Describes pipeline concepts
- Describes recommended practices for building a pipeline
- Walks through a an example of how to take an existing script and turn it into a pipeline.

Next steps: After this PR is merged, it would be great to add examples for building advanced pipelines.

build-pipeline.md was generated from build-pipeline.ipynb using scripts/nb_to_md.py. Please add review comments to the markdown version of the file.

### CODE REVIEW

File: content/en/docs/components/pipelines/sdk/build-pipeline.md [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+*   The component's interface, its inputs and outputs.
+*   The component's implementation, the container image and the command to
+    execute.
+*   The component's metadata, such as the name and description of the
+    component.
+
+You can build components by [defining a component specification for a
+containerized application][component-dev], or you can [use the Kubeflow
+Pipelines SDK to generate a component specification for a Python
+function][python-function-component]. You can also [reuse prebuilt components
```
Comment (8bitmp3) [inline content/en/docs/components/pipelines/sdk/build-pipeline.md:105]: Do you know whether we should use "prebuilt" or "pre-built"? I'd use the former, which is what you've got, but I'm not 100% sure it's the correct way, as the GCloud Docs use both https://cloud.google.com/s/results?q=prebuilt

File: content/en/docs/components/pipelines/sdk/build-pipeline.md [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+that you need to build your pipeline.
+
+*   Components should have a single responsibility. Having a single
+    responsibility makes it easier to test and reuse a component. For example,
+    if you have a component that loads data you can reuse that for similar
+    tasks that load data. If you have a component that loads and transforms
+    a dataset, the component can be less useful since you can use it only when
+    you need to load and transform that dataset. 
+
+*   Reuse components when possible. Kubeflow Pipelines provides [components for
```
Comment (8bitmp3) [inline content/en/docs/components/pipelines/sdk/build-pipeline.md:158]: Just wanted to double check that "Kubeflow Pipelines" is "it" and not "they" (hence, "provides" and not "provide").

File: content/en/docs/components/pipelines/sdk/build-pipeline.md [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+        from a website. The second step extracts the CSV files from a zipped
+        tar file and merges them into a single file. 
+        
+        This approach has a few benefits:
+        
+        *   You can reuse the [Web Download component][web-download-component]
+            to implement the first step.
+        *   Each step has a single responsibility, which makes the components
+            easier to reuse.
+        *   The the zipped tar file is an artifact of the first pipeline step.
```
Comment (8bitmp3) [inline content/en/docs/components/pipelines/sdk/build-pipeline.md:301]: Suggestion (nit):

```suggestion
        *   The zipped tar file is an artifact of the first pipeline step.
```

File: content/en/docs/components/pipelines/sdk/build-pipeline.md [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+            This means that you can examine this artifact when debugging
+            pipelines that use this component.
+    
+    This example implements a two-step pipeline.
+
+### Build your pipeline components
+
+        
+1.  Build you pipeline components. This example modifies the initial script to
```
Comment (8bitmp3) [inline content/en/docs/components/pipelines/sdk/build-pipeline.md:310]: Suggestion:

```suggestion
1.  Build your pipeline components. This example modifies the initial script to
```

File: content/en/docs/components/pipelines/sdk/build-pipeline.md [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+    your component's code as a Docker container image and define the component
+    using a ComponentSpec.
+    
+    In this case, the following modifications were required to the original
+    function.
+
+    *   The file download logic was removed. The path to the zipped tar file
+        is passed as an argument to this function.
+    *   The import statements were moved inside of the function. Python
+        function-based component require standalone Python functions. This
```
Comment (8bitmp3) [inline content/en/docs/components/pipelines/sdk/build-pipeline.md:324]: Suggestion (plural):

```suggestion
        function-based components require standalone Python functions. This
```

**Comment by joeliedtke** [comment]
Note: The links to download the notebook from the markdown file will not work until the PR is merged.

Preview:
https://deploy-preview-2653--competent-brattain-de2d6d.netlify.app/docs/components/pipelines/sdk/build-pipeline/

File: content/en/docs/components/pipelines/sdk/build-pipeline.md (hunk 1 above)
Comment (joeliedtke) [reply content/en/docs/components/pipelines/sdk/build-pipeline.md:105]: The Google Developer Docs Style Guide recommends using the closed form of compound words and words with prefixes. So, it looks like our preference is correct.

https://developers.google.com/style/word-list#compounds

File: content/en/docs/components/pipelines/sdk/build-pipeline.md (hunk 2 above)
Comment (joeliedtke) [reply content/en/docs/components/pipelines/sdk/build-pipeline.md:158]: Yes, in this context "Kubeflow Pipelines" is singular.

File: content/en/docs/components/pipelines/sdk/build-pipeline.md (hunk 3 above)
Comment (joeliedtke) [reply content/en/docs/components/pipelines/sdk/build-pipeline.md:301]: Oops! Thank you for catching this!

File: content/en/docs/components/pipelines/sdk/build-pipeline.md (hunk 4 above)
Comment (joeliedtke) [reply content/en/docs/components/pipelines/sdk/build-pipeline.md:310]: Yikes, thank you for catching this!

File: content/en/docs/components/pipelines/sdk/build-pipeline.md (hunk 5 above)
Comment (joeliedtke) [reply content/en/docs/components/pipelines/sdk/build-pipeline.md:324]: Thanks for catching that!

File: content/en/docs/components/pipelines/sdk/build-pipeline.md (hunk 1 above)
Comment (8bitmp3) [reply content/en/docs/components/pipelines/sdk/build-pipeline.md:105]: Awesome, that's good to know! Thanks @joeliedtke 👍

**Review by RFMVasconcelos — COMMENTED** [review]
This is amazing work @joeliedtke !

**Comment by chensun** [comment]
/lgtm 

Thanks!

File: content/en/docs/components/pipelines/sdk/build-pipeline.md [hunk 6]
```diff
@@ -0,0 +1,465 @@
...
+
+3.  Run the following to print the first few rows of the
+    merged CSV file.
+
+
+```python
+$ head merged_data.csv
+```
+
+4.  Design your pipeline. For example, consider the following pipeline designs.
```
Comment (thedriftofwords) [inline content/en/docs/components/pipelines/sdk/build-pipeline.md:280]: "consider the following pipeline designs:"

**Review by thedriftofwords — APPROVED** [review]
Just one punctuation nit. Otherwise looks great!

File: content/en/docs/components/pipelines/sdk/build-pipeline.md (hunk 6 above)
Comment (joeliedtke) [reply content/en/docs/components/pipelines/sdk/build-pipeline.md:280]: Thanks! I have another PR coming up with some small changes to the doc, so I will include this change in that PR.

**Comment by joeliedtke** [comment]
/unhold

Thanks for your feedback!

**Pull request merged**